ebpf/pprof/pprof.go

```diff
@@ -26,21 +27,96 @@ var (
 	}
 )
 
+type SampleType uint32
+
+var SampleTypeCpu = SampleType(0)
+var SampleTypeMem = SampleType(1)
+
+type SampleAggregation bool
+
+var (
+	// SampleAggregated mean samples are accumulated in ebpf, no need to dedup these
+	SampleAggregated = SampleAggregation(true)
+	// SampleNotAggregated mean values are not accumulated in ebpf, but streamed to userspace with value=1
+	// TODO make consider aggregating python in ebpf as well
+	SampleNotAggregated = SampleAggregation(false)
+)
+
+type CollectProfilesCallback func(sample ProfileSample)
+
+type SamplesCollector interface {
+	CollectProfiles(callback CollectProfilesCallback) error
+}
+
+type ProfileSample struct {
+	Target      *sd.Target
+	Pid         uint32
+	SampleType  SampleType
+	Aggregation SampleAggregation
+	Stack       []string
+	Value       uint64
+	Value2      uint64
+}
+
+type BuildersOptions struct {
+	SampleRate    int64
+	PerPIDProfile bool
+}
+
+type builderHashKey struct {
+	labelsHash uint64
+	pid        uint32
+	sampleType SampleType
+}
+
 type ProfileBuilders struct {
-	Builders   map[uint64]*ProfileBuilder
-	SampleRate int64
+	Builders map[builderHashKey]*ProfileBuilder
+	opt      BuildersOptions
 }
 
-func NewProfileBuilders(sampleRate int64) *ProfileBuilders {
-	return &ProfileBuilders{Builders: make(map[uint64]*ProfileBuilder), SampleRate: sampleRate}
+func NewProfileBuilders(options BuildersOptions) *ProfileBuilders {
+	return &ProfileBuilders{Builders: make(map[builderHashKey]*ProfileBuilder), opt: options}
 }
 
-func (b ProfileBuilders) BuilderForTarget(hash uint64, labels labels.Labels) *ProfileBuilder {
-	res := b.Builders[hash]
+func Collect(builders *ProfileBuilders, collector SamplesCollector) error {
+	return collector.CollectProfiles(func(sample ProfileSample) {
+		builders.AddSample(&sample)
+	})
+}
+
+func (b *ProfileBuilders) AddSample(sample *ProfileSample) {
+	bb := b.BuilderForSample(sample)
+	if sample.Aggregation == SampleAggregated {
+		bb.CreateSample(sample)
+	} else {
+		bb.CreateSampleOrAddValue(sample)
+	}
+}
+
+func (b *ProfileBuilders) BuilderForSample(sample *ProfileSample) *ProfileBuilder {
+	labelsHash, labels := sample.Target.Labels()
+
+	k := builderHashKey{labelsHash: labelsHash, sampleType: sample.SampleType}
+	if b.opt.PerPIDProfile {
+		k.pid = sample.Pid
+	}
+	res := b.Builders[k]
 	if res != nil {
 		return res
 	}
 
+	var sampleType []*profile.ValueType
+	var periodType *profile.ValueType
+	var period int64
+	if sample.SampleType == SampleTypeCpu {
+		sampleType = []*profile.ValueType{{Type: "cpu", Unit: "nanoseconds"}}
+		periodType = &profile.ValueType{Type: "cpu", Unit: "nanoseconds"}
+		period = time.Second.Nanoseconds() / b.opt.SampleRate
+	} else {
+		sampleType = []*profile.ValueType{{Type: "alloc_objects", Unit: "count"}, {Type: "alloc_space", Unit: "bytes"}}
+		periodType = &profile.ValueType{Type: "space", Unit: "bytes"}
+		period = 512 * 1024 // todo
```

Comment:
?

Reply:
https://github.com/google/pprof/blob/main/proto/profile.proto#L85-L86 Not sure what it should be for memory 🤷 we don't really use that I think anyway.